config/ogdl/src/main/java/org/apache/shiro/config/ogdl/ReflectionBuilder.java

```diff
@@ -290,7 +293,9 @@ private String parseBeanId(String lhs) {
         }
 
         //SHIRO-413: init method must be called for constructed objects that are Initializable
-        LifecycleUtils.init(objects.values());
+        //SHIRO-778: onInit method on AuthenticatingRealm is called twice
+        objects.keySet().stream().filter(key -> !Optional.ofNullable(kvPairs).orElseGet(HashMap::new).keySet().contains(key))
```

Comment:
I see another issue with this line. You are using `HashMap::new`. Let's just pretend you are following my comment to remove the overhead of instantiating the map for each key. We might have a line like this:

`Map kvPairsToCheck = Optional.ofNullable(kvPairs).orElseGet(HashMap::new)`.

If you expect the devs to modify this map in a later commit, please do not use `HashMap::new` which is prone to `ConcurrentModificationException`. A lot of apache projects prohibit it's use via checkstyle for this reason.

If you do not expect devs to modify this map in later commits, you might as well use [`Collections::emptyMap`](https://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#emptyMap--), which will hint the dev that this map is going to stay empty (because `Collections` will usually return unmodifiable maps). This way, if someone tries to add content to this map to a later release, he/she will receive an `UnsupportedOperationException` which will make them think about why this was unmodifiable in the first place.
As this only kicks in if the original map `kvPairs` is empty, this seems a sane choice.

Then, `Collections::emptyMap` might receive performance updates in future releases, while `HashMap::new` is less likely to.

Reply:
Oh, btw, the javadoc states: 
> Implementations of this method need not create a separate Map object for each call. Using this method is likely to have comparable cost to using the like-named field. (Unlike this method, the field does not provide type safety.)

Reply:
Oh, and of course, instead of calling 
`Optional.ofNullable(kvPairs).orElseGet(HashMap::new).keySet()`
when refactoring into it's own line you could also use this:
`Set knownKeys = Optional.ofNullable(kvPairs).map(Map::keySet).orElseGet(Collections::emptySet)`.

I mean, the filter is about the set containing a key, right? Or if you avoid mapping into a new collection, you can use containsKey instead: 
`.filter(key -> !map.containsKey(key))`.
That's probably even better because it avoids creation of a new collection and is even shorter to read and easier to understand on first sight.

Reply:
Let me make some polish ;)